### PR TITLE
fix: cast to Set if target type is Set

### PR DESCRIPTION
**Pull request by gagoman**
Fixes https://github.com/dart-lang/json_serializable/issues/54

### CODE REVIEW

File: json_serializable/test/src/_json_serializable_test_input.dart [hunk 1]
```diff
@@ -120,6 +120,21 @@ class FinalFieldsNotSetInCtor {
   FinalFieldsNotSetInCtor();
 }
 
+@ShouldGenerate(r'''
+SetSupport _$SetSupportFromJson(Map<String, dynamic> json) {
+  return SetSupport((json['values'] as List)?.map((e) => e as int)?.toSet());
```
Comment (gagoman) [inline json_serializable/test/src/_json_serializable_test_input.dart]: I was thinking to modify first cast (`as List`) to `as Iterable` but realized that `dart:convert` returns list anyway

File: json_serializable/test/src/_json_serializable_test_input.dart [hunk 2]
```diff
@@ -120,6 +120,21 @@ class FinalFieldsNotSetInCtor {
...
+}
+
+Map<String, dynamic> _$SetSupportToJson(SetSupport instance) =>
+    <String, dynamic>{'values': instance.values?.toList()};
```
Comment (gagoman) [inline json_serializable/test/src/_json_serializable_test_input.dart]: I have not changed return type cast (from `toList` to `toSet`) as `Set` is not supported by default by `dart:convert`

**Comment by gagoman** [comment]
Any other tests I should add?

File: json_serializable/test/src/_json_serializable_test_input.dart (hunk 1 above)
Comment (kevmoo) [reply json_serializable/test/src/_json_serializable_test_input.dart]: yup

File: json_serializable/test/src/_json_serializable_test_input.dart (hunk 2 above)
Comment (kevmoo) [reply json_serializable/test/src/_json_serializable_test_input.dart]: yup

**Comment by gagoman** [comment]
Done

**Comment by gagoman** [comment]
What are the cases? I can work on them.

**Comment by kevmoo** [comment]
> What are the cases? I can work on them.

https://github.com/dart-lang/json_serializable/pull/303/files#r215043811

**Pull request merged**